fendermint/testing/materializer/src/validation.rs

```diff
@@ -0,0 +1,382 @@
+        Ok(subnet_name.clone())
+    }
+
+    fn existing_deployment(
+        &mut self,
+        subnet_name: &SubnetName,
+        gateway: H160,
+        registry: H160,
+    ) -> anyhow::Result<VDeployment> {
+        self.ensure_contains(subnet_name)?;
+
+        if gateway == registry {
+            bail!("gateway and registry addresses are the same in {subnet_name:?}: {gateway} == {registry}");
+        }
+
+        Ok(subnet_name.clone())
+    }
+
+    fn default_deployment(&mut self, subnet_name: &SubnetName) -> anyhow::Result<VDeployment> {
+        self.ensure_contains(subnet_name)?;
+        Ok(subnet_name.clone())
+    }
+
+    fn create_root_genesis<'a>(
+        &mut self,
+        subnet_name: &SubnetName,
+        validators: BTreeMap<&'a VAccount, Collateral>,
+        balances: BTreeMap<&'a VAccount, Balance>,
+    ) -> anyhow::Result<VGenesis> {
+        self.ensure_contains(subnet_name)?;
+        let tn = self.network()?;
+
+        if validators.is_empty() {
+            bail!("validators of {subnet_name:?} cannot be empty");
+        }
+
+        let root_balances = self.balances.entry(tn.root()).or_default();
+
+        for (n, b) in balances {
+            let balance = root_balances.entry(n.clone()).or_default();
+            *balance = b.0;
+        }
+
+        Ok(())
+    }
+
+    async fn create_node<'s, 'a>(
+        &'s mut self,
+        node_name: &NodeName,
+        _node_config: NodeConfig<'a, ValidationMaterials>,
+    ) -> anyhow::Result<VNode>
+    where
+        's: 'a,
+    {
+        self.ensure_contains(node_name)?;
+        Ok(node_name.clone())
+    }
+
+    async fn start_node<'s, 'a>(
+        &'s mut self,
+        _node: &'a VNode,
+        _seed_nodes: &'a [&'a VNode],
+    ) -> anyhow::Result<()> {
+        Ok(())
+    }
+
+    async fn create_subnet<'s, 'a>(
+        &'s mut self,
+        _parent_submit_config: &SubmitConfig<'a, ValidationMaterials>,
+        subnet_name: &SubnetName,
+        subnet_config: SubnetConfig<'a, ValidationMaterials>,
+    ) -> anyhow::Result<VSubnet>
+    where
+        's: 'a,
+    {
+        self.ensure_contains(subnet_name)?;
+        // Check that the submitter has balance on the parent subnet to create the child.
+        let parent = parent_name(subnet_name)?;
+        self.ensure_balance(&parent, subnet_config.creator)?;
+        // Insert child subnet balances entry.
+        self.balances
+            .insert(subnet_name.clone(), Default::default());
+        Ok(subnet_name.clone())
+    }
+
+    async fn fund_subnet<'s, 'a>(
+        &'s mut self,
+        _parent_submit_config: &SubmitConfig<'a, ValidationMaterials>,
+        account: &'a VAccount,
+        subnet: &'a VSubnet,
+        amount: TokenAmount,
+        reference: Option<ResourceHash>,
+    ) -> anyhow::Result<()>
+    where
+        's: 'a,
+    {
+        // Debit parent balance; Credit child balance
+        self.fund_from_parent(subnet, account, amount, true)?;
```

Comment:
I'm probably confusing terminology but are we not moving funds from parent to child? So aren't we crediting the parent and debiting the child?

Reply:
I might have been using it wrong. The +/- depends on which side of the balance sheet you are on: https://accountingplay.com/debits-and-credits/

If I'm holding money in the bank, I'm a liability to them, so credit increases my balance, debit decreases. 
If I took at loan from the bank, I'm an asset to them, so credit decreases and debit increases. 

I thought these are more like the "I keep money in the subnet", the subnet owes me money the same way a bank owes it for my deposits, so I thought when I get tokens on a subnet, they are "credited to me".

Reply:
ok wasn't sure if that comment was outdated or not but that makes sense from that perspective